File: passwords.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <openssl/aes.h>
#include <openssl/rand.h>
#include <openssl/sha.h>

// The AES key should be exactly 32 bytes (256 bits)
#define AES_KEY_SIZE 32

void generate_key(unsigned char *key) {
    if (RAND_bytes(key, AES_KEY_SIZE) != 1) {
        fprintf(stderr, "Error generating random key\n");
        exit(1);
    }
}

void encrypt_password(const char *password, const unsigned char *key, unsigned char *output) {
    AES_KEY aesKey;
    AES_set_encrypt_key(key, AES_KEY_SIZE * 8, &aesKey);
    AES_encrypt((const unsigned char *)password, output, &aesKey);
}

void decrypt_password(const unsigned char *input, const unsigned char *key, char *output) {
    AES_KEY aesKey;
    AES_set_decrypt_key(key, AES_KEY_SIZE * 8, &aesKey);
    AES_decrypt(input, (unsigned char *)output, &aesKey);
}

void create_hash(const char *password, const unsigned char *salt, unsigned char *output) {
    SHA256_CTX sha256;
    SHA256_Init(&sha256);
    SHA256_Update(&sha256, salt, AES_KEY_SIZE);
    SHA256_Update(&sha256, password, strlen(password));
    SHA256_Final(output, &sha256);
}

int verify_master_password(const unsigned char *stored_hash, const char *input_password, const unsigned char *salt) {
    unsigned char input_hash[AES_KEY_SIZE];
    create_hash(input_password, salt, input_hash);
    return memcmp(stored_hash, input_hash, AES_KEY_SIZE) == 0;
}

int main() {
    unsigned char master_key[AES_KEY_SIZE];
    generate_key(master_key);

    unsigned char salt[AES_KEY_SIZE];
    RAND_bytes(salt, AES_KEY_SIZE);

    unsigned char stored_hash[AES_KEY_SIZE];
    create_hash("master_password", salt, stored_hash);

    char input_password[256];

    while (1) {
        printf("Enter the master password: ");
        scanf("%s", input_password);

        if (verify_master_password(stored_hash, input_password, salt)) {
            printf("Access granted.\n");
            break;
        } else {
            printf("Access denied. Please try again.\n");
        }
    }

    char service[256];
    char username[256];
    char password[256];
    unsigned char password_key[AES_KEY_SIZE];
    char encrypted_password[256];

    while (1) {
        printf("\nOptions:\n");
        printf("1. Store a new password\n");
        printf("2. Retrieve a stored password\n");
        printf("3. Quit\n");
        printf("Select an option: ");

        int choice;
        scanf("%d", &choice);

        switch (choice) {
            case 1:
                printf("Enter the service or website name: ");
                scanf("%s", service);
                printf("Enter your username: ");
                scanf("%s", username);
                printf("Generate a new secure password? (y/n): ");

                char generate_new_password;
                scanf(" %c", &generate_new_password);

                if (generate_new_password == 'y' || generate_new_password == 'Y') {
                    // Generate a secure random password
                    for (int i = 0; i < AES_KEY_SIZE; i++) {
                        password[i] = (char)(rand() % 256);
                    }
                } else {
                    printf("Enter your password: ");
                    scanf("%s", password);
                }

                generate_key(password_key);
                encrypt_password(password, password_key, encrypted_password);

                FILE *password_file = fopen("passwords.txt", "a");
                fprintf(password_file, "%s %s %s\n", service, username, encrypted_password);
                fclose(password_file);

                printf("Password for %s stored securely.\n", service);
                break;

            case 2:
                printf("Enter the service or website name: ");
                scanf("%s", service);
                printf("Enter your username: ");
                scanf("%s", username);

                FILE *password_file = fopen("passwords.txt", "r");
                char stored_service[256];
                char stored_username[256];
                char stored_encrypted_password[256];
                int found = 0;

                while (fscanf(password_file, "%s %s %s", stored_service, stored_username, stored_encrypted_password) != EOF) {
                    if (strcmp(service, stored_service) == 0 && strcmp(username, stored_username) == 0) {
                        found = 1;

                        decrypt_password(stored_encrypted_password, password_key, password);
                        printf("Password for %s: %s\n", service, password);
                        break;
                    }
                }

                fclose(password_file);

                if (!found) {
                    printf("Password not found.\n");
                }

                break;

            case 3:
                printf("Goodbye!\n");
                return 0;

            default:
                printf("Invalid choice. Please select a valid option.\n");
                break;
        }
    }
}

File: passwords.py
import os
import base64
import getpass
import hashlib
import secrets
from cryptography.fernet import Fernet

# Generate or load a secret key
def generate_key(key_file="secret.key"):
    if not os.path.isfile(key_file):
        key = Fernet.generate_key()
        with open(key_file, "wb") as keyfile:
            keyfile.write(key)

    with open(key_file, "rb") as keyfile:
        return keyfile.read()

# Encrypt a password and return the ciphertext
def encrypt_password(key, password):
    cipher_suite = Fernet(key)
    ciphertext = cipher_suite.encrypt(password.encode())
    return base64.urlsafe_b64encode(ciphertext).decode()

# Decrypt a ciphertext and return the original password
def decrypt_password(key, ciphertext):
    cipher_suite = Fernet(key)
    ciphertext = base64.urlsafe_b64decode(ciphertext.encode())
    password = cipher_suite.decrypt(ciphertext).decode()
    return password

# Generate a secure random password
def generate_secure_password(length=16):
    chars = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789!@#$%^&*()-_"
    return ''.join(secrets.choice(chars) for _ in range(length))

# Create a salted hash of the master password
def create_hash(master_password, salt):
    hasher = hashlib.sha256()
    hasher.update(salt.encode())
    hasher.update(master_password.encode())
    return hasher.hexdigest()

# Verify the master password
def verify_master_password(stored_hash, input_password, salt):
    input_hash = create_hash(input_password, salt)
    return stored_hash == input_hash

# Main password manager function
def password_manager():
    print("Password Manager")

    key = generate_key()

    while True:
        print("\nOptions:")
        print("1. Store a new password")
        print("2. Retrieve a stored password")
        print("3. Change master password")
        print("4. Quit")
        choice = input("Select an option: ")

        if choice == "1":
            service = input("Enter the service or website name: ")
            username = input("Enter your username: )
            
            # Generate a secure random password
            password = generate_secure_password()
            encrypted_password = encrypt_password(key, password)

            with open("passwords.txt", "a") as password_file:
                password_file.write(f"{service} {username} {encrypted_password}\n")
            print(f"Generated and stored a secure password for {service}.")
        elif choice == "2":
            service = input("Enter the service or website name: ")
            username = input("Enter your username: ")

            found = False
            with open("passwords.txt", "r") as password_file:
                for line in password_file:
                    parts = line.split()
                    if len(parts) == 3 and parts[0] == service and parts[1] == username:
                        decrypted_password = decrypt_password(key, parts[2])
                        print(f"Password for {service}: {decrypted_password}")
                        found = True
                        break

            if not found:
                print("Password not found.")
        elif choice == "3":
            # Change master password
            new_master_password = getpass.getpass("Enter the new master password: ")
            salt = os.urandom(16)
            new_hash = create_hash(new_master_password, salt)
            with open("hash.txt", "w") as hash_file:
                hash_file.write(new_hash)
            print("Master password changed.")
        elif choice == "4":
            print("Goodbye!")
            break
        else:
            print("Invalid choice. Please select a valid option.")

if __name__ == "__main__":
    # Check if the master password exists
    if not os.path.isfile("hash.txt"):
        master_password = getpass.getpass("Create a master password: ")
        salt = os.urandom(16)
        password_hash = create_hash(master_password, salt)
        with open("hash.txt", "w") as hash_file:
            hash_file.write(password_hash)

    # Verify the master password
    while True:
        input_password = getpass.getpass("Enter the master password: ")
        with open("hash.txt", "r") as hash_file:
            stored_hash = hash_file.read()
        if verify_master_password(stored_hash, input_password, salt.decode()):
            print("Access granted.")
            password_manager()
            break
        else:
            print("Access denied. Please try again.")
